File: memg.py
#!/usr/bin/env python
import socket
import threading
import sys

CACHE = {}


def main():
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(("127.0.0.1", 11211))
    sock.listen(1)

    if '--single' in sys.argv:
        conn, _ = sock.accept()
        handle_con(conn)
    else:
        while 1:
            conn, _ = sock.accept()
            thread = threading.Thread(target=handle_con, args=(conn,))
            thread.start()

    sock.shutdown(socket.SHUT_RDWR)
    sock.close()


def handle_con(conn):

    try:
        # python 3
        # Disable universal new lines for python 2 compatibility
        sockfile = conn.makefile(newline="", mode="rw")
    except TypeError:
        # python 2
        sockfile = conn.makefile(mode="rw")

    while True:

        line = sockfile.readline()
        if line == "":
            conn.close()
            break

        parts = line.split()
        cmd = parts[0]

        if cmd == "get":
            key = parts[1]

            try:
                val = CACHE[key]
                output(sockfile, "VALUE %s 0 %d\r\n" % (key, len(val)))
                output(sockfile, val + "\r\n")
            except KeyError:
                pass
            output(sockfile, "END\r\n")

        elif cmd == "set":
            key = parts[1]
            #exp = parts[2]
            #flags = parts[3]
            length = int(parts[4])
            val = sockfile.read(length + 2)[:length]
            CACHE[key] = val

            output(sockfile, "STORED\r\n")

        sockfile.flush()


def output(sockfile, string):
    """Actually write to socket"""
    sockfile.write(string)


if __name__ == "__main__":
    main()

File: memg.js
/* Create the cache object */
var CACHE = {};

var handle_data = function (data, sock) {

    "use strict";

    var parts = data.split(' ');
    var cmd = parts[0], key = parts[1], val;

    switch(cmd) {
        
        case "get":
            key = key.split('\r\n')[0];
            val = CACHE[key];
            if (val) {
                var msg = "VALUE " + key + " 0 " + val.length + "\r\n";
                msg += val + "\r\n";
                sock.write(msg);
            }
            sock.write("END\r\n");
            break;

        case "set":
            var tmp = parts[4].split('\r\n');
            if (tmp[1]) {
                var length = +tmp[0];
                CACHE[key] = tmp[1].slice(0, length);
                sock.write("STORED\r\n");
            };
            break;
    };
};

var handle_conn = function (sock) {

    "use strict";

    var data;
    
    sock.on('data', function(chunk) {

        data = chunk.toString()

        if (data) handle_data(data, sock);

    });

    sock.on('end', function () {

        sock.writeHead(200);
        sock.end();

    });

    if (oneshot) process.exit(1);

};

var net = require('net');

/* Are we running in one-shot mode? */
var oneshot = (process.argv[2] === '--single') ? true : false;

/* Create an async network wrapper */
var server = net.createServer(function(sock) {
    
    handle_conn(sock);

});

server.listen(11211);

